### PR TITLE
Add max_retries, replace_resolv, and timeout for slow internet connections

### PR DESCRIPTION
**Pull request by esotericpig**
# Description:

Possible related issues: #2677 #3226 #1701 #585

In short, my internet sucks, which means, I have to call `gem install <some gem>` and `gem update` multiple times until they finally work.

By adding a `max_retries` and `timeout` setting, through my testing, I now only need to call these commands once.

I think this is important for people that live in countries/cities with poor quality of internet (especially developing countries).

## What was the end-user or developer problem that led to this PR?

When running `gem install <some gem>` or `gem update` with bad/slow internet, you'll constantly get this error:

`Failed to open TCP connection to rubygems.org:443 (getaddrinfo: Name or service not known) (SocketError)`

because the DNS resolver times out, or you'll get general timeout errors (Open/Read/SSL connect timeout errors).

## What is your fix for the problem, implemented in this PR?

By default, nothing is set/changed for users.

The user must change their `~/.gemrc` file:

```yaml
:max_retries: 3
:replace_resolv: true
:timeout: 300
```

`max_retries` must be an integer.

`timeout` is the number of seconds, which can be `nil`, a float, or an int. It sets the Open, Read, SSL, and Continue timeouts (via [Net::HTTP](https://ruby-doc.org/stdlib-2.7.0/libdoc/net/http/rdoc/Net/HTTP.html)).

`replace_resolv` will do:

```ruby
require 'resolv'
require 'resolv-replace'
```

In my testing, this didn't improve anything for me, but many Stackoverflow answers cite this as a way to improve DNS issues in Ruby with slow internet. I would recommend not setting this, and instead just using this:

```yaml
:max_retries: 3
:timeout: 300
```

Here are the default values:

```yaml
:max_retries: 0
:replace_resolv: false
:timeout: null # nil
```

______________

# Tasks:

- [x] Describe the problem / feature
- [x] Write tests
- [x] Write code to solve the problem
- [x] Get code review from coworkers / friends

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

### CODE REVIEW

**Comment by simi** [comment]
@esotericpig looks amazing! I have never heard of `resolv-replace`. I'll take a look what it actually do.

If I understand it well, it doesn't change anything by default and you need to enable it in `.gemrc`. :heavy_check_mark: 

:thinking: Later we can also set default retry to non-zero.

**Comment by esotericpig** [comment]
Hi, thanks. I don't know why the tests fail on windows-bundler. Is that due to my code or something else? The failing tests are cryptic to me.

**Comment by deivid-rodriguez** [comment]
@esotericpig It's [a known random test failure](https://github.com/rubygems/rubygems/issues/3448), don't worry about it :+1:

File: lib/rubygems/remote_fetcher.rb [hunk 1]
```diff
@@ -92,6 +114,10 @@ def initialize(proxy=nil, dns=nil, headers={})
     @cert_files = Gem::Request.get_cert_files
 
     @headers = headers
+
+    @max_retries = max_retries.to_i() # Must be an int
```
Comment (bronzdoc) [inline lib/rubygems/remote_fetcher.rb:118]: Not need to use parens at the end of `to_i`

File: lib/rubygems/config_file.rb [hunk 2]
```diff
@@ -204,6 +241,15 @@ def initialize(args)
     @disable_default_gem_server  = @hash[:disable_default_gem_server]  if @hash.key? :disable_default_gem_server
     @sources                     = @hash[:sources]                     if @hash.key? :sources
     @cert_expiration_length_days = @hash[:cert_expiration_length_days] if @hash.key? :cert_expiration_length_days
+    @timeout                     = @hash[:timeout]                     if @hash.key? :timeout
+
+    if @hash.key? :max_retries
+      @max_retries = @hash[:max_retries].to_i() # Must be an int
```
Comment (bronzdoc) [inline lib/rubygems/config_file.rb:247]: No need for parens at the end of `to_i`

File: lib/rubygems/request.rb [hunk 3]
```diff
@@ -27,14 +33,27 @@ def self.proxy_uri(proxy) # :nodoc:
...
+
     @uri = uri
     @request_class = request_class
     @last_modified = last_modified
     @requests = Hash.new 0
     @user_agent = user_agent
 
     @connection_pool = pool
+
+    @max_retries = max_retries.to_i() # Must be an int
```
Comment (bronzdoc) [inline lib/rubygems/request.rb:55]: Not need for parens at the end of `to_i`

File: lib/rubygems/request.rb [hunk 4]
```diff
@@ -191,8 +210,24 @@ def self.get_proxy_from_env(scheme = 'http')
   def perform_request(request) # :nodoc:
     connection = connection_for @uri
 
+    if @max_retries > 0
+      connection.max_retries = @max_retries
+
+      verbose "using max retries of #{@max_retries}"
+    end
+
+    if !@timeout.nil?()
```
Comment (bronzdoc) [inline lib/rubygems/request.rb:219]: Not need for parens at the end of `nil?`

File: lib/rubygems/remote_fetcher.rb (hunk 1 above)
Comment (esotericpig) [reply lib/rubygems/remote_fetcher.rb]: Okay, I changed it. I know it's not necessary, just stylistic choice. If possible, this should be added to rubocop I think, if this is rubygem's style convention.

**Comment by esotericpig** [comment]
simi mentioned that I should change the default value to `nil`. Is this necessary?

There are other values in `config_file.rb` that don't default to `nil`, so I thought it was okay.

If it's necessary, I'll change the default value to `nil`.

File: lib/rubygems/remote_fetcher.rb (hunk 1 above)
Comment (bronzdoc) [reply lib/rubygems/remote_fetcher.rb]: Agree. Thanks for making the changes and sorry for the hassle.

David just added a cop for this https://github.com/rubygems/rubygems/pull/3682

**Comment by deivid-rodriguez** [comment]
This sounds like it could be related to #2662 and might not be needed if we introduce that PR? Can you try that PR and see if your connection behaviour improves?

**Comment by deivid-rodriguez** [comment]
So, I should assume you don't have ipv6?

> Hi thanks. I don't think falling back to IPv4 will fix it, as per the issue, the errors were related to rubygems.org being unreachable, but I never got an error stating it was unreachable. Also, it would eventually download/update after multiple tries on my end.

@sonalkr132 will know best but this sounds very similar to the ipv6 problems people are having.

In any case, I'm not opposed to introducing this (in particular `:timeout` and `:max_retries`, not sure about `:replace_resolve`) I was just curious whether you had ipv6 and whether trying out the linked PR made things better for you.

**Comment by sonalkr132** [comment]
> getaddrinfo: Name or service not known

This doesn't seem like IPv6 issue to me. As the user mentioned, the inability to resolve IP addresses for rubygems.org can occur when they are on a slow network.
IPv6 connection issue is either because of misconfigured IPv6 routes on the user's part or because of misbehaving Fastly PoP servers. `Name or service not known` means the client was not able to resolve rubygems.org IPs.

**Comment by deivid-rodriguez** [comment]
Ah, ok. I thought this case could also be due to IPv6 issues on the client in case there's a IPv6 address configured as the DNS server.

**Comment by deivid-rodriguez** [comment]
If that was the case though, the linked PR wouldn't fix it since it doesn't implement Happy Eyeballs on the DNS resolution itself.

**Comment by sonalkr132** [comment]
Our DNS provider is separate from the CDN provider. I am yet to come across a case where there was a connection issue because the DNS query for the `AAAA` record failed but the `A` record worked.

**Comment by deivid-rodriguez** [comment]
Ok, that's ruled out then! I was just talking from what I found out from reading the RFC, but I definitely lack hands on experience with this issue.

**Comment by deivid-rodriguez** [comment]
> I am yet to come across a case where there was a connection issue because the DNS query for the AAAA record failed but the A record worked.

Technically (if I'm understanding the theory of this right), both DNS queries would fail in the (potential) case I'm talking about here, since both queries would be sent to a IPv6 host. The issue would be using the IPv6 name server without fallback, instead of falling back to a IPv4 name server in case there are issues when resolving over IPv6.

In any case, I totally trust you that this is not an issue in practice, just wanted to clarify what I mean. It would be this section of the RFC: https://tools.ietf.org/html/rfc8305#section-3.1.

**Comment by sonalkr132** [comment]
>  both queries would be sent to a IPv6 host. The issue would be using the IPv6 name server without fallback, instead of falling back to a IPv4 name server in case there are issues when resolving over IPv6.

yes, this is technically possible. 
In my experience, it is very uncommon for anyone to use IPv6 nameserver on their host. An IPv4 nameserver (ex: 8.8.8.8) can respond to both `AAAA` and `A` queries.

**Comment by deivid-rodriguez** [comment]
>  In my experience, it is very uncommon for anyone to use IPv6 nameserver on their host. An IPv4 nameserver (ex: 8.8.8.8) can respond to both AAAA and A queries.

Makes sense, yeah. The little adoption of IPv6 so far seems a bit like a chicken and egg problem to me: people don't support it because it gives trouble, and it gives trouble because people don't support it :laughing:.

**Comment by esotericpig** [comment]
It's been 4 years. Closing my pull request.

**Comment by simi** [comment]
@esotericpig by any chance, have your internet connection situation improved?

**Comment by esotericpig** [comment]
I moved to another place, so it's fine now. However, I do think it'd be good to have these or similar changes to help people who may live in an area with poor internet or far away from the RubyGems servers.

**Pull request closed without merging**